src/sharing/SharingScreen.js

```diff
@@ -0,0 +1,467 @@
+          fullName={user.full_name}
+          onPress={() => {}}
+        />,
+      );
+    });
+    return preview;
+  };
+  handleSend = async () => {
+    const { users, message } = this.state;
+    let messageToSend = message;
+    const { auth, share } = this.props;
+    const to = JSON.stringify(users.map(user => user.user_id));
+    try {
+      showToast('Sending Message...');
+
+      if (share.type === 'image' || share.type === 'file') {
+        const uri = share.type === 'image' ? share.sharedImageUri : share.sharedFileUri;
+        const fileName = uri.split('/').pop();
+        const response = await uploadFile(auth, uri, fileName);
+        messageToSend += `\n[${fileName}](${response.uri})`;
+      }
+      await sendMessage(auth, { content: messageToSend, type: 'private', to });
+    } catch (err) {
+      showToast('Failed to send message');
+      this.finishShare();
+      return;
+    }
+    showToast('Message sent');
+    this.finishShare();
+  };
+  render() {
+    const { message, choosingRecipients } = this.state;
+
+    if (choosingRecipients) {
+      return (
+        <Modal>
+          <ChooseRecipientsScreen onComplete={this.handleChooseRecipients} />
+        </Modal>
+      );
+    }
+
+    const { share } = this.props;
+    let sharePreview = null;
+    if (share.type === 'image') {
+      sharePreview = (
+        <Image
+          source={{ uri: share.sharedImageUri }}
+          width={200}
+          height={200}
+          style={styles.imagePreview}
+        />
+      );
+    }
+
+    return (
+      <>
+        <ScrollView style={styles.wrapper} keyboardShouldPersistTaps>
+          <View style={styles.container}>{sharePreview}</View>
+          <View style={styles.usersPreview}>{this.getUsersPreview()}</View>
+          <ZulipButton
+            onPress={this.toggleChoosingRecipients}
+            style={styles.chooseButton}
+            text="Choose recipients"
+          />
+          <Input
+            value={message}
+            placeholder="Message"
+            onChangeText={this.handleMessageChange}
+            multiline
+          />
+        </ScrollView>
+        <View style={styles.actions}>
+          <ZulipButton onPress={this.finishShare} style={styles.button} secondary text="Cancel" />
+          <ZulipButton
+            style={styles.button}
+            onPress={this.handleSend}
+            text="Send"
+            disabled={!this.isSendButtonEnabled()}
+          />
+        </View>
+      </>
+    );
+  }
+}
+
+type Props = $ReadOnly<{|
+  navigation: NavigationScreenProp<{ params: {| sharedData: SharedData |} }>,
+  subscriptions: Map<number, Subscription>,
+  users: Map<number, User>,
+  dispatch: Dispatch,
+  auth: Auth,
+|}>;
+
+class SharingScreen extends PureComponent<Props> {
+  render() {
+    const { dispatch, navigation, subscriptions, auth } = this.props;
+    const share: SharedData = navigation.state.params.sharedData;
+    const Tabs = createMaterialTopTabNavigator(
```

Comment:
Hmm interesting. I think this call needs to happen at the top level, outside of the `render()` method. It's basically returning a component *type*, i.e. something at the same level as `SharingScreen` itself rather than a particular instance of it. If `render` gets called twice and we get two of these as a result, I expect that will cause the old instance of the old `Tabs` to be discarded and completely replaced by an instance of the new `Tabs`, rather than just have its props updated. Which would mean any UI state is lost, like which tab is selected and what text the user has entered.

I think the `share` value can maybe be passed through like so?
```
          screen: (share) => (
// ...
        <Tabs share={share} />
```
Not sure what the API is that `react-navigation` provides here, but there ought to be *some* way to do it.

For `dispatch`, `subscriptions`, and `auth`, probably simplest to have `StreamScreen` and `PmScreen` each do their own `connect` calls.

Reply:
I've extracted `createMaterialTopTabNavigator` out of `render`.

The snippet you commented does not seem to work though,

Reply:
It looks like in the new revision it's in the constructor. That means we end up getting data (for `share`) from props, in the constructor:
```js
  constructor(props) {
    super(props);
    const { navigation } = this.props;
    const share: SharedData = navigation.state.params.sharedData;

    this.Tabs = createMaterialTopTabNavigator(
      {
        Stream: {
          screen: () => <ShareToStream share={share} />,
```

That isn't good either -- React components don't ordinarily. We should try to write this in a way that fits smoothly with the basic React model of how things work. It's always possible to do something that deviates if it's really necessary, but we should first try to make it work properly, and if we can't then understand why not.

The react-navigation library is meant to be used with React, so I'll be kind of surprised if it really doesn't have a reasonable way to pass through React props in a normal React-compatible way.

Please take a look at the react-navigation docs (https://reactnavigation.org/docs/2.x/ for the version we're currently on), and/or the web, and/or its implementation in your `node_modules/react-navigation`, and try to find the appropriate feature of its API for this. If you're having trouble, ask on #mobile-team and say what you've looked at and tried so far.

Reply:
> Please take a look at the react-navigation docs (https://reactnavigation.org/docs/2.x/ for the version we're currently on), and/or the web, and/or its implementation in your `node_modules/react-navigation`, and try to find the appropriate feature of its API for this. If you're having trouble, ask on #mobile-team and say what you've looked at and tried so far.

There's been some discussion of this, starting from Divyanshu's post [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.23M117.20Handle.20sharing.20from.20other.20apps/near/882122).

Looking back at my reply there, I think `screenProps` is indeed the way React Navigation would like us to handle it. But there's that inconvenient conclusion that, if you do things ["the right way"](https://reactnavigation.org/docs/2.x/common-mistakes/#explicitly-rendering-more-than-one-navigator), only explicitly rendering the root navigator in your own JSX (and no descendent navigators), whatever you pass as `screenProps` is sort of forced to exist in a global context. Unless there's something less well documented, about `screenProps`, that I'm missing.

It would be helpful if we could work out the advantages of following this "render-only-one-navigator" guideline strictly (and bringing our existing code in line with it).

One compromise (and just because it's a compromise doesn't mean it's what we should do) might be to follow their instructions under

> Alternatively, the following would also work because it exposes the `router` static on `AuthenticationScreen` and threads through the `navigation` prop:

That makes it seam like we could wire things up ourselves, and it would give us the means to pass `screenProps` at a more local level.

Reply:
I've just started a #mobile-team topic about this, [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.60screenProps.60.20in.20react-navigation.20v2).

Reply:
Thanks for that suggestion and the demo! And good thought to move to chat for the detailed discussion.

That approach with the `router` static, and passing through the `navigation` prop, sounds good to me.